app/clip_library.py

```
@@ -137,214 +109,338 @@ def _format_timestamp(milliseconds: int) -> str:
-        current["version"] = 1
-        current["created_at"] = _clean(current.get("created_at")) or _clean(created_at) or _utc_timestamp(clip_path.stat().st_mtime)
-        temporary_path = metadata_path.with_name(metadata_path.name + ".tmp")
-        temporary_path.write_text(json.dumps(current, indent=2, sort_keys=True), encoding="utf-8")
-        os.replace(temporary_path, metadata_path)
+    row = _row_for_path(clip_path)
+    allowed = set(TEXT_FIELDS + TIMELINE_FIELDS + ("media_type", "username", "source_key", "clip_number", "clip_title", "clip_title_custom", "created_at"))
+    scalars = {key: value for key, value in updates.items() if key in allowed}
+    with database.transaction(immediate=True) as connection:
+        if scalars:
+            assignments = ", ".join(f"{key} = ?" for key in scalars)
+            connection.execute(f"UPDATE clips SET {assignments}, updated_at = CURRENT_TIMESTAMP WHERE id = ?", [*scalars.values(), row["id"]])
```

Ruff (0.16.2)

[error] Possible SQL injection vector through string-based query construction

(S608)